server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -557,14 +566,28 @@ private static Map<String, Mapper> buildMergedMappers(
                     // If we're merging template mappings when creating an index, then a field definition always
                     // replaces an existing one.
                     if (reason == MergeReason.INDEX_TEMPLATE) {
-                        mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper);
+                        merged = mergeWithMapper;
                     } else {
-                        mergedMappers.put(mergeWithMapper.simpleName(), mergeIntoMapper.merge(mergeWithMapper, objectMergeContext));
+                        merged = mergeIntoMapper.merge(mergeWithMapper, objectMergeContext);
                     }
                 }
+                if (merged != null) {
+                    mergedMappers.put(merged.simpleName(), merged);
+                }
             }
             return Map.copyOf(mergedMappers);
         }
+
+        private static ObjectMapper truncateObjectMapper(MergeReason reason, MapperMergeContext context, ObjectMapper objectMapper) {
+            // there's not enough capacity for the whole object mapper,
+            // so we're just trying to add the shallow object, without it's sub-fields
+            ObjectMapper shallowObjectMapper = objectMapper.withoutMappers();
+            if (context.decrementFieldBudgetIfPossible(shallowObjectMapper.mapperSize())) {
+                // now trying to add the sub-fields one by one via a merge, until we hit the limit
+                return shallowObjectMapper.merge(objectMapper, reason, context);
+            }
+            return null;
```

Comment:
do we test for this specific case where there is no budget to even add the container?

Reply:
I've added a case in `org.elasticsearch.index.mapper.ObjectMapperMergeTests#testMergeWithLimitObjectField` now.